src/main/java/org/springframework/integration/kafka/listener/MetadataStoreOffsetManager.java

```diff
@@ -0,0 +1,207 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.springframework.integration.kafka.listener;
+
+import static com.gs.collections.impl.utility.Iterate.groupBy;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import com.gs.collections.api.RichIterable;
+import com.gs.collections.api.block.function.Function;
+import com.gs.collections.api.block.procedure.Procedure2;
+import com.gs.collections.api.map.MutableMap;
+import com.gs.collections.api.multimap.MutableMultimap;
+import com.gs.collections.impl.map.mutable.ConcurrentHashMap;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.integration.kafka.core.BrokerAddress;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.Result;
+import org.springframework.integration.metadata.MetadataStore;
+import org.springframework.integration.metadata.SimpleMetadataStore;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * An {@link org.springframework.integration.kafka.listener.OffsetManager} that persists offsets into
+ * a {@link org.springframework.integration.metadata.MetadataStore}
+ *
+ * @author Marius Bogoevici
+ */
+public class MetadataStoreOffsetManager implements OffsetManager {
+
+	private final static Log LOG = LogFactory.getLog(MetadataStoreOffsetManager.class);
+
+	private final GetLeaderFunction getLeaderFunction = new GetLeaderFunction();
+
+	private String consumerId = "spring.kafka";
+
+	private MetadataStore metadataStore = new SimpleMetadataStore();
+
+	private ConnectionFactory connectionFactory;
+
+	private long referenceTimestamp = -2;
+
+	private MutableMap<Partition, Long> offsets = new ConcurrentHashMap<Partition, Long>();
+
+	public MetadataStoreOffsetManager(ConnectionFactory connectionFactory) {
+		this(connectionFactory, null);
+	}
+
+	public MetadataStoreOffsetManager(ConnectionFactory connectionFactory, Map<Partition, Long> initialOffsets) {
+		this.connectionFactory = connectionFactory;
+		if (!CollectionUtils.isEmpty(initialOffsets)) {
+			this.offsets = ConcurrentHashMap.newMap(initialOffsets);
+		}
+		else {
+			this.offsets = ConcurrentHashMap.newMap();
+		}
+		if (!CollectionUtils.isEmpty(initialOffsets)) {
+			initializeOffsets(initialOffsets.keySet());
```

Comment:
`metadataStore` can be persistence one, e.g. `RedisMetadataStore`, o. any net connection from ctor is as bad practice, because you do `this.metadataStore.get(asKey(partition))` withing this `initializeOffsets`